<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Perform Local BLAST Searches â€¢ metablastr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><meta property="og:title" content="Perform Local BLAST Searches">
<meta property="og:description" content="This package provides interface functions for users to perform local Basic Local Alignment Search Tool (BLAST) searches. The search results can be efficiently imported into the R session. ">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">metablastr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    <div id="metablastr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#metablastr" class="anchor"></a>metablastr</h1></div>
<div id="an-easy-way-to-perform-local-large-scale-blast-searches-with-r" class="section level3">
<h3 class="hasAnchor">
<a href="#an-easy-way-to-perform-local-large-scale-blast-searches-with-r" class="anchor"></a>An easy way to perform local large-scale BLAST searches with R</h3>
<p>The Basic Local Alignment Search Tool (BLAST) finds regions of sequence similarity between a query and a subject sequence or sequence database.</p>
<p>The <code>metablastr</code> package provides interface functions between R and the standalone (command line tool) version of BLAST. The search report generated by BLAST can then be imported into the R session either as <code>data.frame</code>/<a href="https://cran.r-project.org/web/packages/tibble/index.html">tibble</a> or as <code>PostgresSQL</code> database connection and can be analysed, filtered, or processed with common <a href="http://r4ds.had.co.nz/">R data science</a> tools such as <a href="https://github.com/tidyverse/dplyr">dplyr</a>, <a href="https://github.com/tidyverse/ggplot2">ggplot2</a>, etc.</p>
<p>The <code><a href="reference/read_blast.html">metablastr::read_blast()</a></code> function has a <a href="https://www.postgresql.org/download/">PostgresSQL database</a> backend, which allows users to generate and store very large BLAST reports and still process them in R via the <a href="https://github.com/tidyverse/dplyr">dplyr</a> database notation.</p>
<p>Hence, using <code><a href="reference/read_blast.html">metablastr::read_blast()</a></code> in database mode users can use the familiar <code>dplyr</code> notation to analyze large scale BLAST reports. Additional functions such as <code>blast_nr()</code>, <code>blast_pdb_protein()</code>, <code>blast_refseq_protein()</code>, etc. are designed to perform easy-to-run BLAST searches on a metagenomic scale. Metagenomic data retrieval can be performed via the <a href="https://github.com/HajkD/biomartr">biomartr</a> package. The <code>biomartr</code> and <code>metablastr</code> packages are designed to seamlessly work together. The corresponding BLAST output can then be analyzed using specialized <code>metablastr::filter_blast_*</code> functions.</p>
</div>
<div id="install-metablastr" class="section level3">
<h3 class="hasAnchor">
<a href="#install-metablastr" class="anchor"></a>Install <code>metablastr</code>
</h3>
</div>
<div id="for-linux-users" class="section level3">
<h3 class="hasAnchor">
<a href="#for-linux-users" class="anchor"></a>For Linux Users:</h3>
<p>Please install the <code>libpq-dev</code> library on you linux machine by typing into the terminal:</p>
<pre><code>sudo apt-get install libpq-dev</code></pre>
</div>
<div id="for-all-systems-install-metablastr-by-typing" class="section level3">
<h3 class="hasAnchor">
<a href="#for-all-systems-install-metablastr-by-typing" class="anchor"></a>For all systems install <code>metablastr</code> by typing</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install.packages("devtools")</span>
<span class="co"># install the current version of metablastr on your system</span>
<span class="kw">library</span>(devtools)
<span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"HajkD/metablastr"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>, <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)</code></pre></div>
</div>
<div id="quick-start" class="section level3">
<h3 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(metablastr)
<span class="co"># run blastn (nucleotide to nucleotide search) between example query and subject sequences</span>
blast_test &lt;-<span class="st"> </span><span class="kw"><a href="reference/blast_nucleotide_to_nucleotide.html">blast_nucleotide_to_nucleotide</a></span>(
                 <span class="dt">query   =</span> <span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">'seqs/qry_nn.fa'</span>, <span class="dt">package =</span> <span class="st">'metablastr'</span>),
                 <span class="dt">subject =</span> <span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">'seqs/sbj_nn.fa'</span>, <span class="dt">package =</span> <span class="st">'metablastr'</span>),
                 <span class="dt">output.path =</span> <span class="kw">tempdir</span>(),
                 <span class="dt">db.import  =</span> <span class="ot">FALSE</span>)
                 
<span class="co"># look at BLAST results</span>
blast_test</code></pre></div>
</div>
<div id="discussions-and-bug-reports" class="section level2">
<h2 class="hasAnchor">
<a href="#discussions-and-bug-reports" class="anchor"></a>Discussions and Bug Reports</h2>
<p>I would be very happy to learn more about potential improvements of the concepts and functions provided in this package.</p>
<p>Furthermore, in case you find some bugs, need additional (more flexible) functionality of parts of this package, or want to contribute to this project please let me know:</p>
<p><a href="https://github.com/HajkD/metablastr/issues" class="uri">https://github.com/HajkD/metablastr/issues</a></p>
</div>
<div id="interfaces-implemented-in-metablastr" class="section level2">
<h2 class="hasAnchor">
<a href="#interfaces-implemented-in-metablastr" class="anchor"></a>Interfaces implemented in <code>metablastr</code>:</h2>
<div id="perform-blast-searches" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-blast-searches" class="anchor"></a>Perform BLAST searches</h3>
<ul>
<li>
<code><a href="reference/blast_protein_to_protein.html">blast_protein_to_protein()</a></code>: Perform Protein to Protein BLAST Searches (BLASTP)</li>
<li>
<code><a href="reference/blast_nucleotide_to_nucleotide.html">blast_nucleotide_to_nucleotide()</a></code>: Perform Nucleotide to Nucleotide BLAST Searches (BLASTN)</li>
<li>
<code><a href="reference/blast_nucleotide_to_protein.html">blast_nucleotide_to_protein()</a></code>: Perform Nucleotide to Protein BLAST Searches (BLASTX)</li>
<li>
<code><a href="reference/blast_protein_to_nucleotide.html">blast_protein_to_nucleotide()</a></code>: Perform Protein to Nucleotide BLAST Searches (TBLASTN)</li>
<li>
<code>blast_best_hit()</code>: Retrieve only the best BLAST hit for each query</li>
<li>
<code>blast_best_reciprocal_hit()</code>: Retrieve only the best reciprocal BLAST hit for each query</li>
<li>
<code><a href="reference/read_blast.html">read_blast()</a></code>: Import BLAST output into R session (in memory) or via <code>PostgresSQL</code> database connection.</li>
</ul>
</div>
<div id="blast-against-common-ncbi-databases" class="section level3">
<h3 class="hasAnchor">
<a href="#blast-against-common-ncbi-databases" class="anchor"></a>BLAST against common NCBI databases</h3>
<ul>
<li>
<code>blast_nr()</code>: Perform Protein to Protein BLAST Searches against the <code>NCBI non-redundant database</code>
</li>
<li>
<code>blast_nt()</code>: Perform Nucleotide to Nucleotide BLAST Searches against the <code>NCBI non-redundant database</code>
</li>
<li>
<code>blast_est()</code>: Perform Nucleotide to Nucleotide BLAST Searches against the <code>NCBI expressed sequence tags database</code>
</li>
<li>
<code>blast_pdb_protein()</code>:</li>
<li>
<code>blast_pdb_nucleotide()</code>:</li>
<li>
<code>blast_swissprot()</code>:</li>
<li>
<code>blast_delta()</code>:</li>
<li>
<code>blast_refseq_rna()</code>:</li>
<li>
<code>blast_refseq_gene()</code>:</li>
<li>
<code>blast_refseq_protein()</code>:</li>
</ul>
</div>
<div id="analyze-blast-report" class="section level3">
<h3 class="hasAnchor">
<a href="#analyze-blast-report" class="anchor"></a>Analyze BLAST Report</h3>
<ul>
<li>
<code>filter_blast_</code>:</li>
</ul>
</div>
<div id="navigation-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#navigation-functions" class="anchor"></a>Navigation functions</h3>
<ul>
<li>
<code><a href="reference/list_outformats.html">list_outformats()</a></code>: List available BLAST output formats</li>
</ul>
</div>
</div>
</div>
  </div>

  <div class="col-md-3" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Report a bug at <br><a href="NA">NA</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Hajk-Georg Drost <br><small class="roles"> Author, contributor, maintainer </small> <a href="https://orcid.org/http://orcid.org/0000-0002-1567-306X" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" height="16"></a> </li>
<li>Anna Gogleva <br><small class="roles"> Author, contributor </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Hajk-Georg Drost, Anna Gogleva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
