<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perfrom BLAST Searches Against a Set of Genomes — blast_genomes • metablastr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Perfrom BLAST Searches Against a Set of Genomes — blast_genomes" />

<meta property="og:description" content="This function takes a fasta file containing query sequences
as input and performs BLAST searches of these query sequences against
a set of reference genomes to retrieve corresponding hits in diverse genomes." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metablastr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perfrom BLAST Searches Against a Set of Genomes</h1>
    
    <div class="hidden name"><code>blast_genomes.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function takes a fasta file containing query sequences
as input and performs BLAST searches of these query sequences against
a set of reference genomes to retrieve corresponding hits in diverse genomes.</p>
    
    </div>

    <pre class="usage"><span class='fu'>blast_genomes</span>(<span class='no'>query</span>, <span class='no'>subject_genomes</span>, <span class='kw'>blast_type</span> <span class='kw'>=</span> <span class='st'>"blastn"</span>,
  <span class='kw'>blast_output_path</span> <span class='kw'>=</span> <span class='st'>"blast_output"</span>, <span class='kw'>min_alig_length</span> <span class='kw'>=</span> <span class='fl'>50</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>query</th>
      <td><p>path to input file in fasta format.</p></td>
    </tr>
    <tr>
      <th>subject_genomes</th>
      <td><p>a vector containing file paths to the reference genomes that shall be queried (e.g. file paths returned by <code><a href='http://www.rdocumentation.org/packages/biomartr/topics/meta.retrieval'>meta.retrieval</a></code>).</p></td>
    </tr>
    <tr>
      <th>blast_type</th>
      <td><p>specification of the BLAST type shall be used to perform BLAST searches between query and reference.
Available options are:</p><ul>
<li><p><code>blast_type = "blastn"</code> for  nucleotide to nucleotide BLAST searches (see <code><a href='blast_nucleotide_to_nucleotide.html'>blast_nucleotide_to_nucleotide</a></code> for details)</p></li>
<li><p><code>blast_type = "tblastn"</code> for protein to nucleotide BLAST searches (see <code><a href='blast_protein_to_nucleotide.html'>blast_protein_to_nucleotide</a></code> for details)</p></li>
</ul></td>
    </tr>
    <tr>
      <th>blast_output_path</th>
      <td><p>a path to a folder that will be created to store BLAST output tables for each individual query-genome search.</p></td>
    </tr>
    <tr>
      <th>min_alig_length</th>
      <td><p>minimum alignment length that shall be retained in the result dataset. All hit alignments with smaller
hit alignment length will be removed automatically.</p></td>
    </tr>
    <tr>
      <th>&#8230;</th>
      <td><p>additional arguments passed to <code><a href='blast_nucleotide_to_nucleotide.html'>blast_nucleotide_to_nucleotide</a></code> or <code><a href='blast_protein_to_nucleotide.html'>blast_protein_to_nucleotide</a></code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The <code>blast_genomes</code> function enables users to BLAST specific query sequences against a set of reference genomes
and retrieve the corresponding BLAST output.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#details">Details</a></li>
          </ul>

    <h2>Author</h2>
    
Hajk-Georg Drost

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost, Anna Gogleva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

