<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform Nucleotide to Nucleotide BLAST Searches (BLASTN) — blast_nucleotide_to_nucleotide • metablastr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
<meta property="og:title" content="Perform Nucleotide to Nucleotide BLAST Searches (BLASTN) — blast_nucleotide_to_nucleotide" />

<meta property="og:description" content="Run nucleotide to nucleotide BLAST of reference sequences
against a blast-able database or fasta file." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metablastr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform Nucleotide to Nucleotide BLAST Searches (BLASTN)</h1>
    
    <div class="hidden name"><code>blast_nucleotide_to_nucleotide.Rd</code></div>
    </div>

    
    <p>Run nucleotide to nucleotide BLAST of reference sequences
against a blast-able database or fasta file.</p>
    

    <pre class="usage"><span class='fu'>blast_nucleotide_to_nucleotide</span>(<span class='no'>query</span>, <span class='no'>subject</span>, <span class='kw'>strand</span> <span class='kw'>=</span> <span class='st'>"both"</span>,
  <span class='kw'>output.path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>is.subject.db</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>task</span> <span class='kw'>=</span> <span class='st'>"blastn"</span>,
  <span class='kw'>db.import</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>postgres.user</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>evalue</span> <span class='kw'>=</span> <span class='fl'>0.001</span>,
  <span class='kw'>out.format</span> <span class='kw'>=</span> <span class='st'>"csv"</span>, <span class='kw'>cores</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>max.target.seqs</span> <span class='kw'>=</span> <span class='fl'>500</span>,
  <span class='kw'>db.soft.mask</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>db.hard.mask</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>blast.path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>query</th>
      <td><p>path to input file in fasta format.</p></td>
    </tr>
    <tr>
      <th>subject</th>
      <td><p>path to subject file in fasta format or blast-able database.</p></td>
    </tr>
    <tr>
      <th>strand</th>
      <td><p>Query DNA strand(s) to search against database/subject.
Options are:</p><ul>
<li><p><code>strand = "both"</code> : query against both DNA strands.</p></li>
<li><p><code>strand = "minus"</code> : query against minus DNA strand.</p></li>
<li><p><code>strand = "plus"</code> : query against plus DNA strand.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>output.path</th>
      <td><p>path to folder at which BLAST output table shall be stored. 
Default is <code>output.path = NULL</code> (hence <code>getwd()</code> is used).</p></td>
    </tr>
    <tr>
      <th>is.subject.db</th>
      <td><p>logical specifying whether or not the <code>subject</code> file is a file in fasta format (<code>is.subject.db = FALSE</code>; default)
or a blast-able database that was formatted with <code>makeblastdb</code> (<code>is.subject.db = TRUE</code>).</p></td>
    </tr>
    <tr>
      <th>task</th>
      <td><p>nucleotide search task option. Options are:</p><ul>
<li><p><code>task = "blastn"</code> : Standard nucleotide-nucleotide comparisons (default) - Traditional BLASTN requiring an exact match of 11.</p></li>
<li><p><code>task = "blastn-short"</code> : Optimized nucleotide-nucleotide comparisons for query sequences shorter than 50 nucleotides.</p></li>
<li><p><code>task = "dc-megablast"</code> : Discontiguous megablast used to find more distant (e.g., interspecies) sequences.</p></li>
<li><p><code>task = "megablast"</code> : Traditional megablast used to find very similar (e.g., intraspecies or closely related species) sequences.</p></li>
<li><p><code>task = "rmblastn"</code></p></li>
</ul></td>
    </tr>
    <tr>
      <th>db.import</th>
      <td><p>shall the BLAST output be stored in a PostgresSQL database and shall a connection be established to this database? Default is <code>db.import = FALSE</code>.
In case users wish to to only generate a BLAST output file without importing it to the current R session they can specify <code>db.import = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>postgres.user</th>
      <td><p>when <code>db.import = TRUE</code> and <code>out.format = "postgres"</code> is selected, the BLAST output is imported and stored in a 
PostgresSQL database. In that case, users need to have PostgresSQL installed and initialized on their system. 
Please consult the Installation Vignette for details.</p></td>
    </tr>
    <tr>
      <th>evalue</th>
      <td><p>Expectation value (E) threshold for saving hits (default: <code>evalue = 0.001</code>).</p></td>
    </tr>
    <tr>
      <th>out.format</th>
      <td><p>a character string specifying the format of the file in which the BLAST results shall be stored.
Available options are:</p><ul>
<li><p><code>out.format = "pair"</code> : Pairwise</p></li>
<li><p><code>out.format = "qa.ident"</code> : Query-anchored showing identities</p></li>
<li><p><code>out.format = "qa.nonident"</code> : Query-anchored no identities</p></li>
<li><p><code>out.format = "fq.ident"</code> : Flat query-anchored showing identities</p></li>
<li><p><code>out.format = "fq.nonident"</code> : Flat query-anchored no identities</p></li>
<li><p><code>out.format = "xml"</code> : XML</p></li>
<li><p><code>out.format = "tab"</code> : Tabular separated file</p></li>
<li><p><code>out.format = "tab.comment"</code> : Tabular separated file with comment lines</p></li>
<li><p><code>out.format = "ASN.1.text"</code> : Seqalign (Text ASN.1)</p></li>
<li><p><code>out.format = "ASN.1.binary"</code> : Seqalign (Binary ASN.1)</p></li>
<li><p><code>out.format = "csv"</code> : Comma-separated values</p></li>
<li><p><code>out.format = "ASN.1"</code> : BLAST archive (ASN.1)</p></li>
<li><p><code>out.format = "json.seq.aln"</code> : Seqalign (JSON)</p></li>
<li><p><code>out.format = "json.blast.multi"</code> : Multiple-file BLAST JSON</p></li>
<li><p><code>out.format = "xml2.blast.multi"</code> : Multiple-file BLAST XML2</p></li>
<li><p><code>out.format = "json.blast.single"</code> : Single-file BLAST JSON</p></li>
<li><p><code>out.format = "xml2.blast.single"</code> : Single-file BLAST XML2</p></li>
<li><p><code>out.format = "SAM"</code> : Sequence Alignment/Map (SAM)</p></li>
<li><p><code>out.format = "report"</code> : Organism Report</p></li>
</ul></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>number of cores for parallel BLAST searches.</p></td>
    </tr>
    <tr>
      <th>max.target.seqs</th>
      <td><p>maximum number of aligned sequences that shall be kept.</p></td>
    </tr>
    <tr>
      <th>db.soft.mask</th>
      <td><p>shall low complexity regions be soft masked? Default is <code>db.soft.mask = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>db.hard.mask</th>
      <td><p>shall low complexity regions be hard masked? Default is <code>db.hard.mask = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>blast.path</th>
      <td><p>path to BLAST executables.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='blast_protein_to_protein.html'>blast_protein_to_protein</a></code>, <code><a href='blast_nucleotide_to_protein.html'>blast_nucleotide_to_protein</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='no'>blast_test</span> <span class='kw'>&lt;-</span> <span class='fu'>blast_nucleotide_to_nucleotide</span>(
                 <span class='kw'>query</span>   <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/qry_nn.fa'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'metablastr'</span>),
                 <span class='kw'>subject</span> <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/sbj_nn.fa'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'metablastr'</span>),
                 <span class='kw'>output.path</span> <span class='kw'>=</span> <span class='fu'>tempdir</span>(),
                 <span class='kw'>db.import</span>  <span class='kw'>=</span> <span class='fl'>FALSE</span>)

 <span class='co'># look at results</span>
 <span class='no'>blast_test</span>
<span class='co'># }</span><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Hajk-Georg Drost

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost, Anna Gogleva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

