<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Extract random loci from a set of genomes — extract_random_seqs_from_multiple_genomes • metablastr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Extract random loci from a set of genomes — extract_random_seqs_from_multiple_genomes" />
<meta property="og:description" content="In some cases, users may wish to extract sequences from randomly sampled loci of a particular length from a set of genomes.
This function allows users to specify a number of sequences
of a specified length that shall be randomly sampled from the genome.
The sampling rule is as follows:
For each locus independently sample:
1) choose randomly (equal probability: see sample.int for details) from which of the given chromosomes the locus shall be sampled (replace = TRUE).
2) choose randomly (equal probability: see sample.int for details) from which strand (plus or minus) the locus shall be sampled (replace = TRUE).
3) randomly choose (equal probability: see sample.int the starting position of the locus in the sampled chromosome and strand (replace = TRUE).

" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metablastr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/installation.html">Installing metablastr and all sequence search tools</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Extract random loci from a set of genomes</h1>
    
    <div class="hidden name"><code>extract_random_seqs_from_multiple_genomes.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>In some cases, users may wish to extract sequences from randomly sampled loci of a particular length from a set of genomes.
This function allows users to specify a number of sequences
of a specified length that shall be randomly sampled from the genome.
The sampling rule is as follows:
For each locus independently sample:</p><ul>
<li><p>1) choose randomly (equal probability: see <code><a href='https://rdrr.io/r/base/sample.html'>sample.int</a></code> for details) from which of the given chromosomes the locus shall be sampled (<code>replace = TRUE</code>).</p></li>
<li><p>2) choose randomly (equal probability: see <code><a href='https://rdrr.io/r/base/sample.html'>sample.int</a></code> for details) from which strand (plus or minus) the locus shall be sampled (<code>replace = TRUE</code>).</p></li>
<li><p>3) randomly choose (equal probability: see <code><a href='https://rdrr.io/r/base/sample.html'>sample.int</a></code> the starting position of the locus in the sampled chromosome and strand (<code>replace = TRUE</code>).</p></li>
</ul>

    </div>

    <pre class="usage"><span class='fu'>extract_random_seqs_from_multiple_genomes</span>(
  <span class='no'>sample_size</span>,
  <span class='kw'>replace</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>prob</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>interval_width</span>,
  <span class='no'>subject_genomes</span>,
  <span class='kw'>file_name</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>separated_by_genome</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>update</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>sample_size</th>
      <td><p>a non-negative integer giving the number of loci that shall be sampled.</p></td>
    </tr>
    <tr>
      <th>replace</th>
      <td><p>logical value indicating whether sampling should be with replacement. Default: <code>replace = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>prob</th>
      <td><p>a vector of probability weights for obtaining the elements of the vector being sampled. Default is <code>prob = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>interval_width</th>
      <td><p>the length of the locus that shall be sampled.</p></td>
    </tr>
    <tr>
      <th>subject_genomes</th>
      <td><p>a vector containing file paths to the reference genomes that shall be queried (e.g. file paths returned by <code><a href='https://docs.ropensci.org/biomartr/reference/meta.retrieval.html'>meta.retrieval</a></code>).</p></td>
    </tr>
    <tr>
      <th>file_name</th>
      <td><p>name of the fasta file that stores the BLAST hit sequences. This name will only be used when <code>separated_by_genome = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>separated_by_genome</th>
      <td><p>a logical value indicating whether or not hit sequences from different genomes should be stored in the same
output <code>fasta</code> file <code>separated_by_genome = FALSE</code> (default) or in separate <code>fasta</code> files <code>separated_by_genome = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>update</th>
      <td><p>shall an existing <code>file_name</code> file be overwritten (<code>update = TRUE</code>; Default) or shall blast hit sequences be appended to the existing file (<code>update = FALSE</code>)?</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>a folder path in which corresponding <code>fasta</code> output files shall be stored.</p></td>
    </tr>
    </table>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
    </ul>

    <h2>Author</h2>
    <p>Hajk-Georg Drost</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost, Anna Gogleva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


