<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ metablastr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Function reference" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metablastr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/installation.html">Installing metablastr and all sequence search tools</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="blast2bed.html">blast2bed()</a></code> </p>
        </td>
        <td><p>Export BLAST output table to BED file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast_best_hit.html">blast_best_hit()</a></code> </p>
        </td>
        <td><p>Retrieve only the best BLAST hit for each query</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast_best_reciprocal_hit.html">blast_best_reciprocal_hit()</a></code> </p>
        </td>
        <td><p>Retrieve only the best reciprocal BLAST hit for each query</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast_genomes.html">blast_genomes()</a></code> </p>
        </td>
        <td><p>Perfrom BLAST Searches Against a Set of Genomes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast_nucleotide_to_nucleotide.html">blast_nucleotide_to_nucleotide()</a></code> </p>
        </td>
        <td><p>Perform Nucleotide to Nucleotide BLAST Searches (BLASTN)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast_nucleotide_to_protein.html">blast_nucleotide_to_protein()</a></code> </p>
        </td>
        <td><p>Perform Nucleotide to Protein BLAST Searches (BLASTX)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast_protein_to_nucleotide.html">blast_protein_to_nucleotide()</a></code> </p>
        </td>
        <td><p>Perform Protein to Nucleotide BLAST Searches (TBLASTN)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast_protein_to_protein.html">blast_protein_to_protein()</a></code> </p>
        </td>
        <td><p>Perform Protein to Protein BLAST Searches (BLASTP)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast_repbase.html">blast_repbase()</a></code> </p>
        </td>
        <td><p>Query the RepBase to annotate putative LTRs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="blast_rpsblast.html">blast_rpsblast()</a></code> </p>
        </td>
        <td><p>Perform Reverse PSI-BLAST searches (rpsblast)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detect_homologs_cds_to_cds.html">detect_homologs_cds_to_cds()</a></code> </p>
        </td>
        <td><p>Perform CDS to CDS BLAST Searches against a set of CDS files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detect_homologs_proteome_to_proteome.html">detect_homologs_proteome_to_proteome()</a></code> </p>
        </td>
        <td><p>Perform Proteome to Proteome BLAST Searches against a set of Proteomes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_hit_seqs_from_genomes.html">extract_hit_seqs_from_genomes()</a></code> </p>
        </td>
        <td><p>Extract sequences of BLAST hits in respective genomes and store it as <code>fasta</code> file(s)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_random_seqs_from_genome.html">extract_random_seqs_from_genome()</a></code> </p>
        </td>
        <td><p>Extract random loci from a genome of interest</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_upstream_promotor_seqs.html">extract_upstream_promotor_seqs()</a></code> </p>
        </td>
        <td><p>Retrieve all upstream promotor sequences from a genome</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="file_contains_aa.html">file_contains_aa()</a></code> </p>
        </td>
        <td><p>Helper function to check whether input file contains amino acids as required</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="file_contains_dna.html">file_contains_dna()</a></code> </p>
        </td>
        <td><p>Helper function to check whether input file contains dna as required</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_best_hits.html">filter_best_hits()</a></code> </p>
        </td>
        <td><p>Retrieve the best hits across species from a BLAST table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_homologs_core_set.html">filter_homologs_core_set()</a></code> </p>
        </td>
        <td><p>Retrieve a core set of best hits that is shared across species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_homologs_core_set_n_species.html">filter_homologs_core_set_n_species()</a></code> </p>
        </td>
        <td><p>Retrieve a core set of best hits that is shared across at least N species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gg_blast_hits.html">gg_blast_hits()</a></code> </p>
        </td>
        <td><p>Plot <code>blast_genomes</code> Result</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gg_hits_boxplot.html">gg_hits_boxplot()</a></code> </p>
        </td>
        <td><p>Boxplot of a <code>blast_tbl</code> generated with <code>detect_homologs_genome_to_genome</code> or <code>detect_homologs_proteome_to_proteome</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gg_homologs_one_to_one.html">gg_homologs_one_to_one()</a></code> </p>
        </td>
        <td><p>Plot number of BLAST hits per species as barplot from a <code>tibble</code> generated with <code>detect_homologs_genome_to_genome</code> or <code>detect_homologs_proteome_to_proteome</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gg_pairwise_orthologs_line.html">gg_pairwise_orthologs_line()</a></code> </p>
        </td>
        <td><p>Plot number of pairwise orthologs for multiple species comparisons</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gg_species_dnds_blast_tbl.html">gg_species_dnds_blast_tbl()</a></code> </p>
        </td>
        <td><p>Plot pairwise dN, dS, or dN/dS distributions of orthologous genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gg_species_feature_blast_tbl.html">gg_species_feature_blast_tbl()</a></code> </p>
        </td>
        <td><p>Plot pairwise feature distributions of orthologous genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="list_outformats.html">list_outformats()</a></code> </p>
        </td>
        <td><p>List available BLAST output formats</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="motif_compare.html">motif_compare()</a></code> </p>
        </td>
        <td><p>Count the number of motifs in a set of non-random versus random sequences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="motif_compare_multi.html">motif_compare_multi()</a></code> </p>
        </td>
        <td><p>Count the number of motifs in a set of non-random versus randomly sampled genomic sequences for multiple species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="motif_compare_multi_promotor_seqs.html">motif_compare_multi_promotor_seqs()</a></code> </p>
        </td>
        <td><p>Count the number of motifs in a set of non-random versus randomly drawn gene promotor sequences for multiple species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="motif_count.html">motif_count()</a></code> </p>
        </td>
        <td><p>Count the number of motifs in a set of sequences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="motif_count_matrix.html">motif_count_matrix()</a></code> </p>
        </td>
        <td><p>Count the number of motifs in a set of sequences and generate a motif count matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="motif_enrichment.html">motif_enrichment()</a></code> </p>
        </td>
        <td><p>Test motif enrichment in a set of non-random versus random sequences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="motif_enrichment_multi.html">motif_enrichment_multi()</a></code> </p>
        </td>
        <td><p>Statistical assessment of motif enrichments in a set of non-random versus randomly sampled genomic sequences for multiple species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="motif_enrichment_multi_promotor_seqs.html">motif_enrichment_multi_promotor_seqs()</a></code> </p>
        </td>
        <td><p>Statistical assessment of motif enrichments in a set of non-random versus randomly sampled gene promotor sequences for multiple species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_blast.html">read_blast()</a></code> </p>
        </td>
        <td><p>Import BLAST output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sample_chromosome_intervals.html">sample_chromosome_intervals()</a></code> </p>
        </td>
        <td><p>Helper function to sample random intervals of length <code>interval_width</code> from chromosomes</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost, Anna Gogleva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

